includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
+  curl_close($cURL);
+  if ($response === false) {
+    $messages = array(
+      'message' => t("Unable to receive response from curl_exec() for object @pid"),
+      'message_sub' => array('@pid' => $object->id),
+      'type' => 'watchdog',
+      'severity' => WATCHDOG_ERROR,
+    );
+    $failureMessage['messages'][] = $messages;
+  } else {
+    $headers = substr($response, 0, $header_size);
+    $body = substr($response, $header_size);
+    if (empty($expect) || (isset($expect[$response_code]) && empty($expect[$response_code]))) {
+      return $body;
+    } else if (isset($expect[$response_code])) {
+      $expectation = $expect[$response_code];
+      $json = json_decode($body, true);
+      if (is_array($json)) {
+        if (isset($json[$expectation])) {
+          return $json;
+        } else {
+          $messages = array(
+            'message' => t("For object @pid : JSON key from $endpoint ($expectation) missing . $response"),
+            'message_sub' => array('@pid' => $object->id),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+        }
+      } else {
+        $messages = array(
+          'message' => t("For object @pid : Expected JSON key from $endpoint but could not decode. $response"),
+          'message_sub' => array('@pid' => $object->id),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+      }
+    } else {
+      $messages = array(
+        'message' => t("For object @pid : Expected one of " .implode(',', array_keys($expect))."; got a $response_code"),
+        'message_sub' => array('@pid' => $object->id), 
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+  }
+  return;
+}
+
+/*
+ * Uploads file to Panopto using AWS S3 Uploader
+ * @param $upload_target - url of upload target
+ * @param $file_path - path of object
+ */
+function islandora_video_streaming_panopto_upload_file($upload_target, $file_path) {
+  $element = explode("/", $upload_target);
+  $prefix = array_pop($element);
+  $service_endpoint = implode("/", $element);
+  $bucket = array_pop($element);
+  $file_name = basename($file_path);
+  $object_key = $prefix."/".$file_name;
+  
+  $s3Params = array(
+      'endpoint' => $service_endpoint,
+      'region'  => 'us-east-1',
+      'version' => '2006-03-01',
+      'credentials' => [
+        'key'    => 'dummy',
+        'secret' => 'dummy'
+      ]
+  );
+  // Create an S3Client
+  $s3Client = new S3Client($s3Params);
+
+  $source = fopen($file_path, 'rb');
+  $uploader = new ObjectUploader(
+    $s3Client,
+    $bucket,
+    $object_key,
+    $source
+  );
+
+  do {
+    try {
+      $result = $uploader->upload();
+    } catch (Exception $ex) {
+      $messages = array(
```

Comment:
In the sample code, Cai-Cian looked for a specific exception (`MultipartUploadException`) and switched from an `ObjectUploader` to a `MultipartUploader` in the `catch`:
https://github.com/ulsdevteam/Panopto_uploader/blob/181c318a6bab857f7aa66fa34d883a15c9310b41/demo.php#L174-L178

What is your rationale for dropping that fallback?

Reply:
I previously had Cai-Cian's code and every time an exception was caught localhost would stop connecting and log "getState()" does not exist.

Reply:
Exception [`MultipartUploadException` does define `getState()`](https://docs.aws.amazon.com/aws-sdk-php/v3/api/class-Aws.Exception.MultipartUploadException.html), so that message is curious.  Is it possible you were catching a different exception?

The documentation for `ObjectUploader` suggests it should use ["a PutObject command or a multipart upload as appropriate."](https://docs.aws.amazon.com/aws-sdk-php/v3/api/class-Aws.S3.ObjectUploader.html), so I'm not terribly concerned about the removal of the fallback, especially since I don't see parallel code in the [Panopto-sourced python example](https://github.com/Panopto/upload-python-sample/blob/master/simplest/panopto_uploader.py).

Reply:
No I am sure the error log listed that multiple of times for different objects.

Reply:
Actually it might've been an error on my end. I had used Exception instead of MultipartUploadException.

Reply:
Yes, capturing any `Exception` rather than just a `MultipartUploadException` would cause this.